presto-memory-context/src/main/java/com/facebook/presto/memory/MemoryTrackingContext.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.memory;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This class is used to track memory usage at all levels (operator, driver, pipeline, etc.).
+ *
+ * At every level we have three aggregate and three local memory contexts. The local memory contexts
+ * track the allocations in the current level while the aggregate memory contexts aggregate the memory
+ * allocated by the children levels and the current level.
+ *
+ * The reason we have local memory contexts at every level is that not all the
+ * allocations are done by the children levels (e.g., at the pipeline level exchange clients
+ * can do system allocations directly, see the ExchangeOperator, another example is the buffers
+ * doing system allocations at the task context level, etc.).
+ *
+ * As another example, at the pipeline level there will be system allocations initiated by the operator context
+ * and there will be system allocations initiated by the exchange clients (local allocations). All these system
+ * allocations will be visible in the systemReservedAggregateMemoryContext.
+ *
+ * To perform local allocations clients should call localUserMemoryContext()/localSystemMemoryContext()
+ * and get a reference to the local memory contexts. The other child-originated allocations will go through
+ * reserveUserMemory()/reserveRevocableMemory()/setSystemMemory() methods.
+ */
+@ThreadSafe
+public class MemoryTrackingContext
+{
+    private final AggregatedMemoryContext userReservedAggregateMemoryContext;
+    private final AggregatedMemoryContext userRevocableAggregateMemoryContext;
+    private final AggregatedMemoryContext systemReservedAggregateMemoryContext;
```

Comment:
The names in this class is a bit too tense.

Use user/revocable/system instead of userReserved/userRevocable/systemReserved
Use a single update function instead of a pair of reserve/free functions
Use getXXXMemory instead of reservedXXXMemory